Fix #7752: Menu not collapsed when clicking on webviews

EstherPerelman opened the pull request:
Signed-off-by: Esther Perelman esther.perelman@sap.com

#### What it does

Fixes #7752 

Solution description: 'phosphor' is listening to widgets 'mousedown' event and so knows that the menu should be closed, All the other widgets except the webview - dispatching a 'mousedown' event automatically on user click but on iframe click - the node not dispatching 'mousedown' event - so we need to do it manually. (on WebviewMessageChannels.didFocus - dispatching node 'mousedown').

#### How to test

1. Open some webview.
2. Click on the 'file' top menu item/ Click on the 'settings' icon on the bottom left.
3. Click somewhere on the webview and see the the menu is collapsed.

Please note: Because of #8572 - after this fix the menu sometimes closes while navigating because the webview getting the focus, Fixing #8572 will solve it too.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto reviewed (commented):
@EstherPerelman thank you for the contribution :+1: 
I believe the 'ECA Check' is currently failing due to your sign-off, it does not seem to be the correct format:

```
Signed-off-by: Esther Perelman esther.perelman@sap.com
```

rather it should be:

```
Signed-off-by: Esther Perelman <esther.perelman@sap.com>
```

I verified that the ECA is successfully completed.

marechal-p commented:
@EstherPerelman Can you explain why it was required to add `this.node.dispatchEvent(new MouseEvent('mousedown'));`? It works, but I don't clearly see why. LGTM otherwise.

vince-fugnitto commented:
> @EstherPerelman Can you explain why it was required to add `this.node.dispatchEvent(new MouseEvent('mousedown'));`? It works, but I don't clearly see why. LGTM otherwise.

I think a description like Paul mentioned in the pull-request description will help reviewers understand why the changes are necessary and why this approach was taken.

amiramw commented:
> I think a description like Paul mentioned in the pull-request description will help reviewers understand why the changes are necessary and why this approach was taken.

Probably also a comment in the code

EstherPerelman commented:
> @EstherPerelman thank you for the contribution 👍
> I believe the 'ECA Check' is currently failing due to your sign-off, it does not seem to be the correct format:
> 
> ```
> Signed-off-by: Esther Perelman esther.perelman@sap.com
> ```
> 
> rather it should be:
> 
> ```
> Signed-off-by: Esther Perelman <esther.perelman@sap.com>
> ```
> 
> I verified that the ECA is successfully completed.

Thanks, Fixed it.

> @EstherPerelman Can you explain why it was required to add `this.node.dispatchEvent(new MouseEvent('mousedown'));`? It works, but I don't clearly see why. LGTM otherwise.

Added Solution description.


> > I think a description like Paul mentioned in the pull-request description will help reviewers understand why the changes are necessary and why this approach was taken.
> 
> Probably also a comment in the code

Added comment, Hope it's enough...

vince-fugnitto reviewed (commented):
I'm not sure about the tradeoff of menus being dismissed randomly when the webview steals focus versus menus not being dismissed when a webview is clicked. As a user, the random dismisal of menus feels more intrusive than the menu not being able to dismiss when a webview is clicked.

I think we should try to investigate why this occurs (I see you added a note about it).

EstherPerelman commented:
Hi all, I moved the code to a better place and expanded the comment.

EstherPerelman commented:
> I'm not sure about the tradeoff of menus being dismissed randomly when the webview steals focus versus menus not being dismissed when a webview is clicked. As a user, the random dismisal of menus feels more intrusive than the menu not being able to dismiss when a webview is clicked.
> 
> I think we should try to investigate why this occurs (I see you added a note about it).

I'm checking and trying to resolve this..

EstherPerelman commented:
> > > > I'm not sure about the tradeoff of menus being dismissed randomly when the webview steals focus versus menus not being dismissed when a webview is clicked. As a user, the random dismisal of menus feels more intrusive than the menu not being able to dismiss when a webview is clicked.
> > > > I think we should try to investigate why this occurs (I see you added a note about it).
> > > 
> > > 
> > > I'm checking and trying to resolve this..
> > 
> > 
> > Solved by #8576
> 
> @EstherPerelman should we close this pull-request in favor of #8576?

No, This PR is needed for closing open menu on webview focus,
#8576 Fixes the random dismissal of menus on menu navigation with open webview.

vince-fugnitto commented:
> No, This PR is needed for closing open menu on webview focus,

I'd think that they should go together, or at least in the same pull-request (two commits if absolutely necessary).

> #8576 Fixes the random dismissal of menus on menu navigation with open webview.

This bug or breaking behavior is introduced by this pull-request and that is my reasoning as to why they should be together.
The bug in `master` is simply the unnecessary focusing of webviews, but the random dismissal of menus is introduced in this pull-request. We should not introduce a new bug to fix an old one unless we decide on a complete solution (ex: both commits together).

EstherPerelman commented:
> > No, This PR is needed for closing open menu on webview focus,
> 
> I'd think that they should go together, or at least in the same pull-request (two commits if absolutely necessary).
> 
> > #8576 Fixes the random dismissal of menus on menu navigation with open webview.
> 
> This bug or breaking behavior is introduced by this pull-request and that is my reasoning as to why they should be together.
> The bug in `master` is simply the unnecessary focusing of webviews, but the random dismissal of menus is introduced in this pull-request. We should not introduce a new bug to fix an old one unless we decide on a complete solution (ex: both commits together).

So what in the end - Which PR should we keep (with the complete solution) ? current or #8576

vince-fugnitto commented:
> So what in the end - Which PR should we keep (with the complete solution) ? current or #8576

Whichever you prefer, I just suggested they should be together.
Together they give us the complete solution without introducing additional issues, and we can more easily review as well.

EstherPerelman commented:
> > So what in the end - Which PR should we keep (with the complete solution) ? current or #8576
> 
> Whichever you prefer, I just suggested they should be together.
> Together they give us the complete solution without introducing additional issues, and we can more easily review as well.

I also think so.. updated #8576 you can close current PR

vince-fugnitto commented:
Closing in favor of #8576.
